### PR TITLE
feat(middleware/csrf): TrustedOrigins using https://*.example.com style subdomains

### PR DESCRIPTION
**Pull request by sixcolors**
### PR Description:

This PR updates the `TrustedOrigins` subdomain matching style to match the changes made in the CORS middleware. Here's a summary of the changes:

#### Changes Made:

- **Documentation Update**:
  - Updated the documentation in `docs/api/middleware/csrf.md` to reflect the new subdomain matching style for `TrustedOrigins`.
  - Updated the `TrustedOrigins` section to include examples and cautionary notes regarding subdomain matching.
- **Code Changes**:
  - Modified the `csrf.go` file to parse and handle `TrustedOrigins` with subdomain matching.
  - Added a new `helpers.go` file containing helper functions for normalizing origins and matching subdomains.
  - Updated corresponding tests in `csrf_test.go` and added new tests in `helpers_test.go` to ensure the correct behavior of subdomain matching.

#### Notes for Review:

- **Documentation Clarity**: Please review the updated documentation in `docs/api/middleware/csrf.md` to ensure clarity and correctness.
- **Code Logic**: Review the changes in the `csrf.go` file to confirm that the handling of `TrustedOrigins` with subdomain matching aligns with our requirements.
- **Test Coverage**: Verify that the tests in `csrf_test.go` cover various scenarios related to subdomain matching and ensure they provide sufficient coverage.
- **Performance Impact**: Consider any potential performance impact introduced by the changes in the `helpers.go` file and ensure it's acceptable.

Kindly review and provide feedback. Thank you!

## Summary by CodeRabbit

- **Refactor**
	- Improved CSRF protection by enhancing origin and referer matching to better handle subdomains.
- **Tests**
	- Updated and added tests to ensure robust CSRF protection, including handling of wildcard domains and invalid origins.
- **Documentation**
	- Updated `TrustedOrigins` configuration in CSRF middleware to support subdomain matching with wildcard characters. Added examples and security guidance in the documentation.

### CODE REVIEW

File: middleware/csrf/csrf_test.go [hunk 1]
```diff
@@ -767,8 +755,8 @@
 		TrustedOrigins: []string{
 			"http://safe.example.com",
 			"https://safe.example.com",
-			"http://.domain-1.com",
-			"https://.domain-1.com",
+			"http://*.domain-1.com",
+			"https://*.domain-1.com",
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/csrf/csrf_test.go:759]: sure make a commitable suggestion to append those to an existing test.

File: middleware/csrf/helpers.go [hunk 2]
```diff
@@ -11,3 +13,45 @@ func compareTokens(a, b []byte) bool {
...
+}
+
+type subdomain struct {
+	prefix string
+	suffix string
+}
+
+func (s subdomain) match(o string) bool {
+	return len(o) >= len(s.prefix)+len(s.suffix) && strings.HasPrefix(o, s.prefix) && strings.HasSuffix(o, s.suffix)
+}
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/csrf/helpers.go:57]: normalizeOrigin() which is called when building trustedSubOrigins and originMatchesHost() and refererMatchesHost() all do this already.

File: middleware/csrf/helpers_test.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+		if valid != tc.expectedValid {
+			t.Errorf("Expected origin '%s' to be valid: %v, but got: %v", tc.origin, tc.expectedValid, valid)
+		}
+
+		if normalizedOrigin != tc.expectedOrigin {
+			t.Errorf("Expected normalized origin '%s' for origin '%s', but got: '%s'", tc.expectedOrigin, tc.origin, normalizedOrigin)
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/csrf/helpers_test.go:53]: yeah go for it.

File: middleware/csrf/helpers_test.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := tt.sub.match(tt.origin)
+			assert.Equal(t, tt.expected, got, "subdomain.match()")
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/csrf/helpers_test.go:113]: no we handle the case issues elsewhere in the code.

File: middleware/csrf/csrf.go [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 import (
 	"errors"
+	"log"
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/csrf/csrf.go:5]: We think it's important to panic if a security middleware has a misconfiguration.

File: middleware/csrf/csrf.go [hunk 6]
```diff
@@ -273,23 +295,31 @@
...
 // returns an error if the referer header is not present or is invalid
 // returns nil if the referer header is valid
-func refererMatchesHost(c fiber.Ctx, trustedOrigins []string) error {
-	referer := c.Get(fiber.HeaderReferer)
+func refererMatchesHost(c fiber.Ctx, trustedOrigins []string, trustedSubOrigins []subdomain) error {
+	referer := strings.ToLower(c.Get(fiber.HeaderReferer))
+
 	if referer == "" {
 		return ErrRefererNotFound
 	}
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/csrf/csrf.go:325]: You should check that again.

File: middleware/csrf/helpers.go [hunk 7]
```diff
@@ -11,3 +13,45 @@ func compareTokens(a, b []byte) bool {
...
+
+	// Validate there is a host present. The presence of a path, query, or fragment components
+	// is checked, but a trailing "/" (indicative of the root) is allowed for the path and will be normalized
+	if parsedOrigin.Host == "" || (parsedOrigin.Path != "" && parsedOrigin.Path != "/") || parsedOrigin.RawQuery != "" || parsedOrigin.Fragment != "" {
+		return false, ""
+	}
+
+	// Normalize the origin by constructing it from the scheme and host.
+	// The path or trailing slash is not included in the normalized origin.
+	return true, strings.ToLower(parsedOrigin.Scheme) + "://" + strings.ToLower(parsedOrigin.Host)
```
Comment (ReneWerner87) [inline middleware/csrf/helpers.go:47]: can you try https://github.com/gofiber/utils/blob/master/strings.go#L8

File: middleware/csrf/csrf.go [hunk 8]
```diff
@@ -273,23 +294,31 @@ func originMatchesHost(c fiber.Ctx, trustedOrigins []string) error {
...
+	return ErrOriginNoMatch
 }
 
 // refererMatchesHost checks that the referer header matches the host header
 // returns an error if the referer header is not present or is invalid
 // returns nil if the referer header is valid
-func refererMatchesHost(c fiber.Ctx, trustedOrigins []string) error {
-	referer := c.Get(fiber.HeaderReferer)
+func refererMatchesHost(c fiber.Ctx, trustedOrigins []string, trustedSubOrigins []subdomain) error {
+	referer := strings.ToLower(c.Get(fiber.HeaderReferer))
```
Comment (ReneWerner87) [inline middleware/csrf/csrf.go:320]: can you try https://github.com/gofiber/utils/blob/master/strings.go#L8

File: middleware/csrf/csrf.go [hunk 9]
```diff
@@ -262,8 +283,8 @@ func isFromCookie(extractor any) bool {
 // originMatchesHost checks that the origin header matches the host header
 // returns an error if the origin header is not present or is invalid
 // returns nil if the origin header is valid
-func originMatchesHost(c fiber.Ctx, trustedOrigins []string) error {
-	origin := c.Get(fiber.HeaderOrigin)
+func originMatchesHost(c fiber.Ctx, trustedOrigins []string, trustedSubOrigins []subdomain) error {
+	origin := strings.ToLower(c.Get(fiber.HeaderOrigin))
```
Comment (ReneWerner87) [inline middleware/csrf/csrf.go:287]: can you try https://github.com/gofiber/utils/blob/master/strings.go#L8

**Review by ReneWerner87 — APPROVED** [review]

**Comment by sixcolors** [comment]
@ReneWerner87 

| Benchmark                          | ns/op (avg) | B/op (avg) | allocs/op (avg) |
|------------------------------------|-------------|------------|-----------------|
|  strings.ToLower()             |             |            |                 |
| Benchmark_Middleware_CSRF_Check-24| 3913.25     | 412        | 9               |
| utils.ToLower()              |             |            |                 |
| Benchmark_Middleware_CSRF_Check-24| 4006.625    | 436        | 10              |

I would note that the Benchmark_Middleware_CSRF_Check is testing a best case scenario for `strings.ToLower()` where the Referer header is already in lowercase and does not contain any non ASCII chars. 

And, the benchmark for CSRF had a bug, I'll push a fix.

File: middleware/csrf/helpers.go (hunk 7 above)
Comment (ReneWerner87) [reply middleware/csrf/helpers.go:47]: standard lib is faster if it is already lower case, which should be the case in most cases

File: middleware/csrf/csrf.go (hunk 9 above)
Comment (ReneWerner87) [reply middleware/csrf/csrf.go:287]: standard lib is faster if it is already lower case, which should be the case in most cases

**Pull request merged**